### ISSUE TITLE
Wrong type inference with record on expression evaluation

### ISSUE THREAD

**Issue by FMorschel** [dart-lang/sdk#56859]
Repro:

```dart
void main() {
  foo();
}

void foo() {
  bar({
    for (var i = 0; i < 10; i++) (i,),
  });
}

void bar(Set<(int,)> things) {
  baz(things);
}

int baz(Set<(int,)> things) => things.fold(0, (a, b) => a + b.$1);
```

Add a breakpoint inside `bar`.
Now add `baz(things)` to `watch` or run this on the debug console and see:

```
evaluateInFrame: (113) Expression compilation error
org-dartlang-debug:synthetic_debug_expression:1:5: Error: The argument type '_Set<dynamic>' can't be assigned to the parameter type 'Set<(int)>'.
 - '_Set' is from 'dart:_compact_hash'.
 - 'Set' is from 'dart:core'.
baz(things)
    ^
```

If you change the expression to `baz(things.cast())` it works. 

I did some experimenting and if you remove the `Record` from this and use only `int` the expression is evaluated correctly.

Could this be related to the problems we talked about in [this issue](https://github.com/Dart-Code/Dart-Code/issues/5294) that surround `Records` not having an element? 

---

CC: @DanTup, @bwilkerson

**Comment by FMorschel** [comment]
Also, could this be related to https://github.com/dart-lang/sdk/issues/56555 in any way?

**Comment by a-siva** [comment]
//cc @jensjoha @johnniwinther

**Comment by bwilkerson** [comment]
> Could this be related to the problems we talked about in this issue that surround `Records` not having an element?

No. The referenced issue is strictly limited to the analyzer. The compilation error you're seeing is from the VM.

**Comment by johnniwinther** [comment]
@jensjoha Can you take a look at this? My guess that this is caused by the VM passing the runtime type `_Set<dynamic>` instead of the static type `Set<(int,)>`.

**Comment by jensjoha** [comment]
The VM currently sends `[dart:_compact_hash, _Set, 1, 1, null]` --- seemingly from 9a023aeae9ec (even with a comment "_Record class is not useful for the CFE.").

I have a cl currently running testing on the try bots making the VM-side and the dart-side to pass the information along (https://dart-review.googlesource.com/c/sdk/+/388841).
